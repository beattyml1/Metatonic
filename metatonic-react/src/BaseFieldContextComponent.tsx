import * as React from "react";
import {BaseEditorModel} from "metatonic-core";
import {AnyTypeParameterType, SchemaField, SchemaType} from "metatonic-core";
import {getUniqueId} from "metatonic-core";
import {ComponentContext} from "metatonic-core";
import {ValidationMessageDetailed} from "metatonic-core";

export abstract class BaseFieldContextComponent<TParams extends BaseEditorModel<any>, TState = {}>
    extends React.Component<TParams, TState> {

    field() {
        return this.props.field;
    }

    uniqueId() {
        return getUniqueId(this.field(), this.props.context);
    }

    renderValidationMessages() {
        return (
            <div className="validation-list">
                <div >
                    {this.validationMessages().map((message, index) => {
                        <div className={`validation-message ${message.severity}`}
                             id={`${this.uniqueId()}-message-${index}`}
                             role="alert" >
                            {message.message}
                        </div>
                    })}
                </div>
            </div>);
    }

    validationMessages(): ValidationMessageDetailed[] {
        return this.props.fieldState.validationMessages;
    }
}