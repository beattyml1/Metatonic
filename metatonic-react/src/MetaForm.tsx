import * as React from "react";
import {} from 'metatonic-redux/'
import {PersistantDataStore} from "metatonic-core";
import {TopLevelMetatonicComponent} from "./TopLevelMetatonicComponent";

export type MetaFormProps= {
    formName?: string,
    recordName: string,
    recordId?: string,
    title?: string,
    dataStore: PersistantDataStore
    afterLoad?: () => void
}

export class MetaForm<T = any> extends TopLevelMetatonicComponent<T, MetaFormProps> {
    constructor(props: MetaFormProps, context?) {
        super(props, context);
        
    }

    render() {
        return (
            <form onSubmit={this.submit}>
                {this.renderEditor()}
                <button type="button" className="secondary">Cancel</button>
                <button type="submit" className="primary">Save</button>
            </form>
        );
    }

    submit() {
        this.store.trySubmit();
    }
    
    componentDidMount() {
        this.init().then(() => {
            if (this.props.afterLoad) {
                this.props.afterLoad
            }
        })
    }

    async init() {
        let resource = this.props.dataStore.records(this.props.recordName);
        let formData = await resource.getOne(this.props.recordId||"new");
        let schema = await resource.schema();
        return await super.init(formData, schema);
    }
}