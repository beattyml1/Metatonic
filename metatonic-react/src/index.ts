export * from './MetaForm'
export * from './MetaEdit'
export * from './Editors/FieldEditor'
export * from './Editors/BaseEditor'
export * from './Editors/BaseRecordEditor'
export * from './BaseFieldContextComponent'
export * from './LabeledFieldContainers/LabelingProviderBaseComponent'
export * from './LabeledFieldContainers/FieldSet'
export * from './LabeledFieldContainers/InputFieldLabelAndContainer'
export * from './Services/ReactEditorService'