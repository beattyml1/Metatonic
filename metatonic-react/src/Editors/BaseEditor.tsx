import * as React from "react";
import {BaseEditorModel} from "metatonic-core";
import {AnyTypeParameterType, SchemaType} from "metatonic-core";
import {BaseFieldContextComponent} from "../BaseFieldContextComponent";

export abstract class BaseEditor
    <TData,
     TType extends AnyTypeParameterType,
     TParams extends BaseEditorModel<TData>,
     TState = {}>
    extends BaseFieldContextComponent<TParams, TState> {

    fieldLocator() {
        return this.props.context.fieldLocator
    }

    store() {
        return this.props.resources.formStore;
    }

    formDispatcher() {
        return this.props.resources.formDispatcher;
    }

    type() {
        return Object.assign({}, this.props.field.type, this.props.field.type.parameters as TType)
    }

    value() {
        return this.props.value;
    }

    notifyChanged(value) {
        this.props.resources.formDispatcher.propertyChanged(this.fieldLocator(), this.value())
    }
}