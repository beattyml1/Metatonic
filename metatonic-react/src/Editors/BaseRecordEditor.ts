import {BaseEditorModel, ComponentContext,SchemaField} from "metatonic-core";
import {RecordSchemaType} from "metatonic-core";
import {createContext} from "metatonic-core";
import {BaseEditor} from "./BaseEditor";

export abstract class BaseRecordEditor <TData extends {}, TProps extends BaseEditorModel<TData>, TState={}>
    extends BaseEditor<TData, RecordSchemaType, TProps, TState> {

    type() {
        return super.type() as RecordSchemaType;
    }

    childProps(name: string) {
        let field = this.type().parameters.fields.find(f => f.name === name);
        return {
            value: this.props.value[name],
            field: field as SchemaField,
            context: createContext(field, this.props.context) as ComponentContext
        }
    }
}