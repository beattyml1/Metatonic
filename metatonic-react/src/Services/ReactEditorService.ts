import {BaseEditor} from "../Editors/BaseEditor";
import * as React from "react";
import {FormSchema} from "metatonic-core";
import {EditorSubContext} from "metatonic-core";
import {
    editorRegistry, multiEditRegistry,
    selectRegistry, EditorResolver
} from "metatonic-core";

export type BaseEditorComponentClass = new () => BaseEditor<any, any, any, any>
export type LabelContainerClass = new () => React.Component<any, any>;
export type BaseRepeaterClass = new () => React.Component<any, any>

export class ReactEditorResolver extends EditorResolver<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass> {
    constructor(protected schema: FormSchema) {
        super(
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                editorRegistry, schema, false),
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                selectRegistry,schema, true),
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                multiEditRegistry,schema, false)
        );
    }
}
