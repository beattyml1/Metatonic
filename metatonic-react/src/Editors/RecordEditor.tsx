import * as React from "react";
import {FieldEditor} from "./FieldEditor";
import {BaseEditorModel, createContext, editorFor, RecordSchemaType, SchemaRecordTypeParameters} from "metatonic-core";
import FieldSet from "../LabeledFieldContainers/FieldSet";
import {BaseRecordEditor} from "./BaseRecordEditor";
import {fieldEditorClasses} from "metatonic-core";
import {getChildFieldProps} from "metatonic-core";

export type RecordParams = BaseEditorModel<any>;
export type RecordMultiParams = BaseEditorModel<any[]>;

@editorFor("Record", FieldSet, {isDefault: true})
export class RecordEditor extends BaseRecordEditor<{[key:string]:any}, BaseEditorModel<RecordSchemaType>, {}> {
    render() {
        let recordType = this.props.field.type.parameters as SchemaRecordTypeParameters;
        let fields = recordType.fields;
        let fieldEditors = fields.map(field =>
            <FieldEditor {...getChildFieldProps(this.props, field)} />);

        return (<div className={fieldEditorClasses(this.field())}>{fieldEditors}</div>)
    }
}



