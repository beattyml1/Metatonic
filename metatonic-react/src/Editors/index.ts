import './BooleanEditor';
import './DateTimeEditor';
import './RecordEditor';
import './QuantityEditor';
import './NumericEditor';
import './TextFieldEditor';