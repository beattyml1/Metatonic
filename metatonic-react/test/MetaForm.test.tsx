import {RecordEditor} from "../src/Editors/RecordEditor";
import renderer = require('react-test-renderer');
import * as React from "react";
import * as Core from 'metatonic-core'
import {SchemaField} from "metatonic-core";
import {SchemaType} from "metatonic-core";
import {recordSchema,getFormSchemaFromJsonObject} from "../../metatonic-core/test/TestUtils";
import {exampleSchema} from "../../metatonic-core/test/TestSchema";
import {getDefaultDataForField} from "metatonic-core";
import {startNewFormStateManager} from "metatonic-core";
import {FormSchema, ObjectDataStorage} from "metatonic-core";
import {ReactEditorResolver} from "../src/Services/ReactEditorService";
import {editorRegistry} from "../../metatonic-core/src/services/EditorRegistry";
import '../src';
import {addUniqueIdsToChildren} from "../../metatonic-core/src/services/IdGeneratorService";
import {Quantity} from "../../metatonic-core/src/Data/Quantity";
import {Decimal} from "../../metatonic-core/src/Data/Decimal";
import {Integer} from "../../metatonic-core/src/Data/Integer";
import {Date} from "../../metatonic-core/src/Data/Date";
import {findField} from "../../metatonic-core/src/services/FieldNavigationHelpers";
import {copyAndSet} from "../../metatonic-core/src/extensions/functional";
import {MetaForm} from '../src/MetaForm'

describe('RecordEditor', () => {
    function createTopField(type: Core.SchemaType) {
        return {
            label: "",
            uiUniqueId: "",
            type: type,
            typeName: type.name
        } as SchemaField;
    }
    it('renders correctly for new data', () => {
        expect.assertions(2);
        const schema = addUniqueIdsToChildren(getFormSchemaFromJsonObject(exampleSchema), '');
        const type = schema.type;
        const field = createTopField(type);
        let formData = getDefaultDataForField(field);

        let store = startNewFormStateManager();
        store.fullReload(formData, schema);
        let editors = new ReactEditorResolver(schema);

        let component;
        let promise = new Promise<any>((resolve, reject) => {
            component = renderer
                .create(
                    <MetaForm recordName='Home' recordId={null} title='Enter Record' dataStore={new ObjectDataStorage({})} afterLoad={resolve} />)
                ;
            expect(component.toJSON()).toMatchSnapshot();
        });
        return promise.then(() =>
            expect(component.toJSON()).toMatchSnapshot());
    });
    it('renders correctly for existing data', () => {

    });
})