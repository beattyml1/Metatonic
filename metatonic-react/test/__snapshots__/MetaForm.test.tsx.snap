// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordEditor renders correctly for new data 1`] = `
<form
  onSubmit={[Function]}
>
  <div />
  <button
    className="secondary"
    type="button"
  >
    Cancel
  </button>
  <button
    className="primary"
    type="submit"
  >
    Save
  </button>
</form>
`;
